### DISCUSSION TITLE
ESP32 S3 and ES8388

### DISCUSSION THREAD

**Discussion by grender007** [pschatzmann/arduino-audio-driver#2] [Q&A] [answered]
Hi!
I've been playing with this new library and the new addition to audio tools (I2SCodecStream), I think I make some progress. However, I'm not getting anything in the serial monitor. Could you give a hint of what I am missing? I based my code on one of the examples. So far, I didn't get any errors or bugs.
---------------------------------------------------------------------------------------------------------------------------------------------------

#include "AudioTools.h"
#include "AudioLibs/I2SCodecStream.h"

AudioInfo info(44100, 1, 16);                       //Sample Rate - Channels - Bits per Sample

DriverPins my_pins;                                 //Driver API for codec
AudioBoard board(&AudioDriverES8388, my_pins);      //Definition for audio board pins and an audio driver
I2SCodecStream out(board);                          //Provide a board for the I2S stream

CsvOutput<int16_t> sout(Serial);                      // ASCII output stream 
StreamCopy copier(sout, out); 

void setup() {
  // Setup logging
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Info);
    LOGLEVEL_AUDIODRIVER = AudioDriverInfo;

  // Setup pins
    //I2C codec pins: SCL, SDA, PORT, (I2C object)
    my_pins.addI2C(CODEC, 4, 3, 0x22, 100000, Wire);
    //I2S pins: MCLK (master clock), BCLK/SCLK (bit or serial clock), WS/LRCLK (word select), SDOUT (data out), SDIN (data in)
    my_pins.addI2S(CODEC, 14, 36, 8, 18, 15);
  
  //Start I2S & CODEC
    Serial.println("starting I2S...");
    CodecConfig cfg;
      //Mic input to adc channel 1
      cfg.adc_input = ADC_INPUT_LINE1;  
      //Set all left format for I2S
      cfg.i2s.fmt  = I2S_LEFT;
      //Bits per sample (16 bits)                     
      cfg.i2s.bits = BIT_LENGTH_16BITS;
      //Sample Rate (44.1 kHz)
      cfg.i2s.rate = RATE_44K;
    board.begin(cfg);

  auto config = out.defaultConfig();
  config.copyFrom(info);
  out.begin(config);
}

void loop() { 
  copier.copy();
}

**Comment by pschatzmann** [comment, marked as answer]
In your sketch the sout begin call is missing.

I suggest to use [this sketch](https://github.com/pschatzmann/arduino-audio-driver/blob/main/examples/audiotools/audiotools-custom-max/audiotools-custom-max.ino), which is much simpler as starting point.

Please also update the audio-tools and the audio-driver library since there were quite a lot of changes lately

**Reply by grender007** [reply]
If I understand correctly, that sketch sends a sine wave to the codec, right? But does the microcontroller receive the signal back after it was digitized in the coded? I'm trying to use the codec as an external ADC for my ESP.

**Comment by pschatzmann** [comment]
No, that's not how things work: the audio signal is in all cases sent out via the i2s interface.
The only difference is, that the codec needs to be initialized first as a precondition that it works.

The I2SCodecStream sets up both: i2s and the codec and when you write data to it, it gets forwarded to or from i2s (of the codec)

The i2s connection can be douplex: you can write to the DAC and read from the ADC of the codec.

If you want a reading sketch you just copy from the I2SCodecStream to the output (e.g. csv)

**Reply by grender007** [reply]
Hi again!
I'm still playing with the code, but unfortunately I'm stuck again. The code is compiling with no errors, but for some reason I'm not getting anything from the I2SStream. In the serial monitor I get this:

[E] I2SESP32V1.h : 95 - size_t audio_tools::I2SDriverESP32V1::readBytes(void*, size_t)
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 0 -> 0 bytes - in 0 hops
[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops

I double checked my physical connections and the pin configurations, and everything seems correct. My code is attached below.

[test.txt](https://github.com/pschatzmann/arduino-audio-driver/files/14328502/test.txt)

**Comment by Marco1971Repo** [comment]
I'm starting to play with this new library too. I thinkis a wonderfull idea, anyway i'm trying the sample: [[examples/audiotools/audiotools-standard](https://github.com/pschatzmann/arduino-audio-driver/tree/main/examples/audiotools/audiotools-standard)](url)
but no sound fromearphone and the log say:

9:56:17.337 -> [D] StreamCopy.h : 97 - copy 
19:56:17.337 -> [D] StreamCopy.h : 177 - available: 0
19:56:17.337 -> [D] StreamCopy.h : 158 - no data 
19:56:17.337 -> [D] StreamCopy.h : 97 - copy 
19:56:17.337 -> [D] StreamCopy.h : 177 - available: 0
19:56:17.337 -> [D] StreamCopy.h : 158 - no data 
19:56:17.337 -> [D] StreamCopy.h : 97 - copy 

I'm doing something wrong or maybe the library is still work in progress?

**Reply by pschatzmann** [reply]
The link is not working. If it is using the GeneratedSoundStream your library version might have the issue that you need to call begin() on that object. In the latest version I was undoing this change because I have also too many examples which are not doing this!

**Comment by Marco1971Repo** [comment]
You are right missing call to begin on GeneratedSoundStream

**Comment by jay6621** [comment]
Hi, Phil , I am also working on exact same setup i.e. esp32 s3 with es8388,
This gives perfect output on serial plotter, just putting it here for information.
I was trying to plot the audio data from Mic on serial plotter.

with Hardware
Board - ESP32S3 Dev Module
Codec - PCB Artist's ES8388 Board

and software
Arduino IDE 2.3.2
ESP32 Board Version  2.0.17
arduino-audio-driver and Audio Tools - Latest version

with below code Based on audiotools-custom-max sketch with few changes -

```
/**
 * @brief We define a custom board with the i2c and i2s pins and output_device a sine
 * with the help of the AudioTools I2SCodecStream
 * @author phil schatzmann
 */

#include "AudioTools.h" // install https://github.com/pschatzmann/arduino-audio-tools
#include "AudioLibs/AudioBoardStream.h"

// I2C
#define SDAPIN               17 // I2C Data,  Adafruit ESP32 S3 3, Sparkfun Thing Plus C 23
#define SCLPIN               18 // I2C Clock, Adafruit ESP32 S3 4, Sparkfun Thing Plus C 22
#define I2CSPEED        100000 // Clock Rate
#define ES8388ADDR        0x10 // Address of ES8388 I2C port

// I2S, your configuration for the ES8388 board
#define MCLKPIN             0 // Master Clock
#define BCLKPIN             11 // Bit Clock
#define WSPIN               12 // Word select
#define DOPIN               10 // This is connected to DI on ES8388 
#define DIPIN               13 // This is connected to DO on ES8388 

AudioInfo                     audio_info(8000, 1, 16);                // sampling rate, # channels, bit depth
DriverPins                    my_pins;                                 // board pins
AudioBoard                    audio_board(AudioDriverES8388, my_pins); // audio board
I2SCodecStream                i2s_in_stream(audio_board);
CsvOutput<int16_t>            csvStream(Serial);
StreamCopy                    copier(csvStream, i2s_in_stream); // copy kit to csvStream
TwoWire                       myWire = TwoWire(0);

void setup() {
  // Setup logging
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  LOGLEVEL_AUDIODRIVER = AudioDriverInfo;
  delay(2000);

  Serial.println("Setup starting...");

  Serial.println("I2C pin ...");
  my_pins.addI2C(PinFunction::CODEC, SCLPIN, SDAPIN, ES8388ADDR, I2CSPEED, myWire);
  Serial.println("I2S pin ...");
  my_pins.addI2S(PinFunction::CODEC, MCLKPIN, BCLKPIN, WSPIN, DOPIN, DIPIN);

  Serial.println("Pins begin ..."); 
  my_pins.begin();

  Serial.println("Board begin ..."); 
  audio_board.begin();

  Serial.println("I2S begin ..."); 
  auto i2s_config = i2s_in_stream.defaultConfig(RX_MODE);
  i2s_config.copyFrom(audio_info);  
  i2s_config.input_device = ADC_INPUT_LINE1;
  i2s_in_stream.begin(i2s_config); // this should apply I2C and I2S configuration
  csvStream.begin(audio_info);
  Serial.println("Setup completed ...");
}

// Arduino loop - copy sound to out
void loop() { copier.copy(); }

```


Thank you for sharing this library.

JAY

**Comment by jay6621** [comment]
Hi @pschatzmann I wanted to connect es8388 to esp32 s3 and I am interested in raw data from i2s,
In [custom-max](https://github.com/pschatzmann/arduino-audio-driver/blob/main/examples/custom/custom-max/custom-max.ino) example 
> /**
>  * @brief We just set up the codec and the i2c pins, so that the wire is
>  * automatically initialized with the indicated pins.
>  * After this you can set up and use i2s
>  * @author phil schatzmann
>  */

So after setting i2c and codec and i2s can we use 
`#include "driver/i2s.h"`
to get data from i2s using

`i2s_read(I2S_NUM_0, &sampleBuffer, sizeof(sampleBuffer), &bytesRead, portMAX_DELAY);`

or do I need to use audio-tools library ?

**Comment by pschatzmann** [comment]
No, you don't need the AudioTools library, but I would strongly recommend to use it.

The I2S API that you mention is obsolete and when you upgrade the ESP32 core to > 3.0, it will not work any more...
You could use the Arduino I2S API, but here as well, there are some differences, so you end up with come code changes.

ps. this library just sets up the codec chip and not I2s: so you will still need to do this yourself if you don't use the AudioTools CodecStream or AudioBoardStream

**Reply by jay6621** [reply]
That's much useful info, So, now I am going with 
`I2SCodecStream` and will use  `addI2S` to set up i2s pins and
by using [readBytes](https://github.com/pschatzmann/arduino-audio-tools/blob/caea8188c6eeba8d79b10251020440bfb4a81342/src/AudioLibs/I2SCodecStream.h#L127C18-L127C27 ) from `I2SCodecStream ` 
can I get raw data or do I need to use some scaler??

**Reply by pschatzmann** [reply]
You get the data that you configured: I would also define the pins there like in [this example](https://github.com/pschatzmann/arduino-audio-driver/blob/main/examples/audiotools/audiotools-custom-min/audiotools-custom-min.ino)

**Comment by jay6621** [comment]
Hi Phil, @pschatzmann I am using [Wire](https://github.com/espressif/arduino-esp32/tree/master/libraries/Wire) for an i2c sensor (BMA400 accelerometer) and for ES8388 driver I am doing,
```
TwoWire myWire = TwoWire(0);
DriverPins my_pins;
my_pins.addI2C(PinFunction::CODEC, SCLPIN, SDAPIN, ES8388ADDR, I2CSPEED, myWire);
```
In this case none of them works i.e. i2c can't find ES8388 and BMA400 , They works fine separately,
can you suggest the right way to use both ??

**Comment by pschatzmann** [comment]
You would need to analyze why this is not working: Activating leg level debug in the Arduino Tools menu might give some hints.
Make sure that none of your processing is calling end() or begin() with some different pin assignements on your wire object.

Or you can just use separate connections with the predefined Wire and Wire1.

**Comment by jay6621** [comment]
Hi, Phil @pschatzmann 
as per your suggestion I did tried to analyze problem,
by enabling sensor (BMA400 Accelerometer)  first it works but ES8388 fails
and vice versa..
after enabling log I got 
```
I2S begin ...
[ 23825][W][Wire.cpp:311] begin(): Bus already started in Master Mode.
Error:   ->p_wire->endTransmission: 5
Error:   ES8388 not found on I2C address 10, check wiring
Error:   AudioDriver::begin::init failed
Error:   ->p_wire->endTransmission: 5
Error:   ->p_wire->endTransmission: 5
Error:   ->p_wire->endTransmission: 5
Error:   ->p_wire->endTransmission: 5
```
This happens only when Already Wire is initialized by other sensor library.

**Reply by pschatzmann** [reply]
You can retrieve the i2c  definition that you have added and set active = false; 
this prevent the setup of the Wire pins.

Or better just do this in your setup
```
TwoWire myWire = TwoWire(0);
PinsI2C i2c;
i2c.set_active = false;
i2c.function= PinFunction::CODEC;
i2c.p_wire = &myWire
DriverPins my_pins;
my_pins.addI2C(i2c);
```

I committed some additinal functionality where this could be achieved as follows
```
TwoWire myWire = TwoWire(0);
my_pins.addI2C(PinFunction::CODEC, myWire);
```
